### PR TITLE
Support GRPC

### PR DESCRIPTION
**Pull request by neokidd**
### Motivation:

Support GRPC. It is based on #383. But managed to use different ways to get proxy.

### Modification:

Comparing to #383, this support of GRPC uses a different approach to make proxy. It hacks byte code to allow extending GPRC interface. In this way, implementation looks more straight forward. Using this support of GRPC is also more naturally like other SOFA supported protocols.

It keeps minimal changes to core-implement. Only important change to core is a small change allowing GRPC proxy to pass interface check. 

It's merged with the latest master. No conflicts. Ready for PR. 

### Result:

Based on #383. Fixes its conflict issues, as a different approach is used.

It adds GRPC support as an extension. Minimal touch to core.

### CODE REVIEW

**Comment by neokidd** [comment]
Looks it because of the maven plugin not downloaded.  I'm fixing that.

**Comment by neokidd** [comment]
Build issue fixed. Please review.

**Comment by neokidd** [comment]
@ujjboy any progress with code review?

File: core/api/src/main/java/com/alipay/sofa/rpc/config/ConsumerConfig.java [hunk 1]
```diff
@@ -258,7 +258,7 @@ public String buildKey() {
         try {
             if (StringUtils.isNotBlank(interfaceId)) {
                 this.proxyClass = ClassUtils.forName(interfaceId);
-                if (!proxyClass.isInterface()) {
+                if (!"grpc".equals(protocol) && !proxyClass.isInterface()) {
```
Comment (ujjboy) [inline core/api/src/main/java/com/alipay/sofa/rpc/config/ConsumerConfig.java]: `grpc` use constants.

File: example/pom.xml [hunk 2]
```diff
@@ -130,7 +147,35 @@
...
         <plugins>
+            <!-- grpc -->
+            <plugin>
+                <groupId>org.xolstice.maven.plugins</groupId>
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.5.1</version>
+                <configuration>
+                    <protocArtifact>com.google.protobuf:protoc:3.5.1-1:exe:${os.detected.classifier}</protocArtifact>
+                    <pluginId>grpc-java</pluginId>
+                    <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.17.0:exe:${os.detected.classifier}</pluginArtifact>
```
Comment (ujjboy) [inline example/pom.xml]: `3.5.1`  and  `1.17.0` use property

File: example/src/main/java/com/alipay/sofa/rpc/grpc/examples/registry/GrpcClientRegistryApplication.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alipay.sofa.rpc.grpc.registry;
+
+import com.alipay.sofa.rpc.config.RegistryConfig;
+import com.alipay.sofa.rpc.config.ConsumerConfig;
+import io.grpc.examples.helloworld.GreeterGrpc;
```
Comment (ujjboy) [inline example/src/main/java/com/alipay/sofa/rpc/grpc/examples/registry/GrpcClientRegistryApplication.java]: format imports.

File: extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+import javassist.util.proxy.ProxyObject;
+import com.alipay.sofa.rpc.invoke.Invoker;
+import com.alipay.sofa.rpc.message.MessageBuilder;
+
+import com.alipay.sofa.rpc.core.exception.RpcErrorType;
+import com.alipay.sofa.rpc.core.exception.SofaRpcException;
+import com.alipay.sofa.rpc.core.exception.SofaRpcRuntimeException;
+
+import com.alipay.sofa.rpc.core.request.SofaRequest;
+import com.alipay.sofa.rpc.core.response.SofaResponse;
```
Comment (ujjboy) [inline extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java]: format imports.

File: extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    /**
+     * 构造函数
+     
+     * @param consumerConfig 服务消费者配置
+     */
+    protected GrpcConsumerBootstrap(ConsumerConfig<T> consumerConfig) {
+        super(consumerConfig);
+        this.interfaceID = consumerConfig.getInterfaceId();
+        String[] segments = this.interfaceID.split("\\.");
+        String lastSegment = segments[segments.length - 1];
```
Comment (ujjboy) [inline extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java]: why not use method `lastIndexOf `

File: core/api/src/main/java/com/alipay/sofa/rpc/config/ConsumerConfig.java (hunk 1 above)
Comment (neokidd) [reply core/api/src/main/java/com/alipay/sofa/rpc/config/ConsumerConfig.java]: Will make change accordingly.

File: example/pom.xml (hunk 2 above)
Comment (neokidd) [reply example/pom.xml]: Will make change accordingly.

File: example/src/main/java/com/alipay/sofa/rpc/grpc/examples/registry/GrpcClientRegistryApplication.java (hunk 3 above)
Comment (neokidd) [reply example/src/main/java/com/alipay/sofa/rpc/grpc/examples/registry/GrpcClientRegistryApplication.java]: Will make change accordingly.

File: extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java (hunk 4 above)
Comment (neokidd) [reply extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java]: Will make change accordingly.

File: extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java (hunk 5 above)
Comment (neokidd) [reply extension-impl/bootstrap-grpc/src/main/java/com/dib/sofa/rpc/bootstrap/grpc/GrpcConsumerBootstrap.java]: Will make change to use method lastIndexOf

**Comment by leizhiyuan** [comment]
@ujjboy @NeGnail @JervyShi Please review this

**Comment by leizhiyuan** [comment]
please change version to 5.6.3-SNAPSHOT

fix the build problem ,we will merge It into 5.6.3

**Comment by neokidd** [comment]
Crazily busy recently, I will try to look into this over this weekend.

**Comment by neokidd** [comment]
> I will continue to review this pr when I have time.

Done with requested changes.

**Comment by neokidd** [comment]
> please change version to 5.6.3-SNAPSHOT
> 
> fix the build problem ,we will merge It into 5.6.3

Done.

**Pull request merged**